core/org.osate.aadl2/src/org/osate/aadl2/instance/impl/ComponentInstanceImpl.java

```diff
@@ -993,6 +983,113 @@ private static boolean isSameOrRefined(Connection firstconn, Connection secondco
 		}
 		return false;
 	}
+
+	private static boolean isSameOrRefined(Feature f1, Feature f2) {
+		if (f1 == f2) {
+			return true;
+		}
+		Feature cfi = f1;
+		while (cfi != null) {
+			if (f2 == cfi) {
+				return true;
+			}
+			cfi = cfi.getRefined();
+		}
+		cfi = f2;
+		while (cfi != null) {
+			if (f1 == cfi) {
+				return true;
+			}
+			cfi = cfi.getRefined();
+		}
+		return false;
+	}
+
+	private static boolean isSameOrRefined(Subcomponent sub1, Subcomponent sub2) {
+		if (sub1 == sub2) {
+			return true;
+		}
+		Subcomponent rsub = sub1;
+		while (rsub != null) {
+			if (sub2 == rsub) {
+				return true;
+			}
+			rsub = rsub.getRefined();
+		}
+		rsub = sub2;
+		while (rsub != null) {
+			if (sub1 == rsub) {
+				return true;
+			}
+			rsub = rsub.getRefined();
+		}
+		return false;
+	}
+
+	private static boolean isSameOrRefined(FlowSpecification fs1, FlowSpecification fs2) {
+		if (fs1 == fs2) {
+			return true;
+		}
+		FlowSpecification rsub = fs1;
+		while (rsub != null) {
+			if (fs2 == rsub) {
+				return true;
+			}
+			rsub = rsub.getRefined();
+		}
+		rsub = fs2;
+		while (rsub != null) {
+			if (fs1 == rsub) {
+				return true;
+			}
+			rsub = rsub.getRefined();
+		}
+		return false;
+	}
+
+	private static boolean isSameOrRefined(EndToEndFlow e2e1, EndToEndFlow e2e2) {
+		if (e2e1 == e2e2) {
+			return true;
+		}
+		EndToEndFlow rsub = e2e1;
+		while (rsub != null) {
+			if (e2e2 == rsub) {
+				return true;
+			}
+			rsub = rsub.getRefined();
+		}
+		rsub = e2e2;
+		while (rsub != null) {
+			if (e2e1 == rsub) {
+				return true;
+			}
+			rsub = rsub.getRefined();
+		}
+		return false;
+	}
+
+//	private static <T> boolean isSameOrRefined(T firstconn, T secondconn, Function<T, T> getRefined) {
```

Comment:
It is a little confusing that a generic version of `isSameOrRefined` is left in, but commented out. Was this supposed to be deleted, but wasn't? Were the other versions of `isSameOrRefined` supposed to call this one? Either this comment out method should be deleted or it should be used.

Reply:
The commented-out generic version was meant to be removed.  I wanted to use it initially to avoid the duplication of the isSameOrRefined method, but it didn't work out.  The problem is the `Function<T, T>` parameter, which is meant to be satisfied by a lambda expression.  But the execution environment for `org.osate.aadl2` is 1.5 because Lutz says otherwise it doesn't build properly.  So I cannot actually pass a lambda at the call sites, and I didn't want to use ugly anonymous classes.